Src/Support/GoogleApis.Tests/Apis/Upload/ImprovedResumableUploadTest.cs

```diff
@@ -0,0 +1,966 @@
+﻿using Google.Apis.Json;
+using Google.Apis.Services;
+using Google.Apis.Upload;
+using Google.Apis.Util;
+using NUnit.Framework;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Apis.Tests.Apis.Upload
+{
+    /// <summary>
+    /// Tests of resumable upload, that uses a real HTTP server.
+    /// </summary>
+    /// <remarks>
+    /// These tests are very flakey on Travis, due to issue #765, so they are not run on Travis.
+    /// When these tests de-flake on travis, replace ResumableUploadTest with this class.
+    /// </remarks>
+    [TestFixture]
+    [Category(IgnoreOnTravis)]
+    class ImprovedResumableUploadTest
+    {
+        public const string IgnoreOnTravis = "IgnoreOnTravis";
+        /// <summary>
+        /// Mock string to upload to the media server. It contains 454 bytes, and in most cases we will use a chunk 
+        /// size of 100. There are 3 spaces on the end of each line because the original carriage return line endings
+        /// caused differences between Windows and Linux test results.
+        /// </summary>
+        static readonly string UploadTestData =
+            "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod   " +
+            "tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris   " +
+            "nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore   " +
+            "eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit   " +
+            "anim id est laborum.";
+        static readonly byte[] uploadTestBytes = Encoding.UTF8.GetBytes(UploadTestData);
+        static readonly int uploadLength = uploadTestBytes.Length;
+
+        /// <summary>
+        /// Stream that doesn't support seeking.
+        /// </summary>
+        private class UnknownSizeMemoryStream : MemoryStream
+        {
+            public UnknownSizeMemoryStream(byte[] buffer) : base(buffer) { }
+            public override bool CanSeek => false;
+            public override long Seek(long offset, SeekOrigin loc)
+            {
+                throw new NotSupportedException();
+            }
+            public override long Position
+            {
+                get { return base.Position; }
+                set { throw new NotSupportedException(); }
+            }
+        }
+
+        /// <summary>
+        /// resumable upload class that allows chunksize to be set for testing.
+        /// </summary>
+        private class MockResumableUpload : ResumableUpload<object>
```

Comment:
Were we going to rename this, given that it's not mocking anything? How about "TestResumableUpload"? Doesn't get the "chunk size" bit, admittedly...